src/main/java/uk/gov/ons/ctp/response/sample/ingest/CsvIngesterBusiness.java

```diff
@@ -26,138 +32,197 @@
+            lineNumber++;
+            try {
+                Pair<BusinessSampleUnit, List<ValidationError>> pair = parseLine(nextLine, unitRefs, lineNumber);
+                List<ValidationError> newErrors = pair.getValue();
+
+                if (newErrors.size() > 0){
+                    errors.addAll(newErrors);
+                } else {
+                    sampleUnitList.add(pair.getKey());
+                }
+            } catch (CTPException e) {
+                String newMessage = String.format("Line %d: %s", csvReader.getRecordsRead(), e.getMessage());
+                throw new CTPException(e.getFault(), e, newMessage);
+            }
         }
-      }
-      SampleSummary sampleSummaryWithCICount = sampleService.saveSample(sampleSummary, sampleUnitList, SampleUnitState.INIT);
-      publishToPartyQueue(sampleUnitList, sampleSummary.getId().toString());
-
-      return sampleSummaryWithCICount;
-  }
-
-  private BusinessSampleUnit parseLine(String[] nextLine, Set<String> unitRefs) throws IllegalAccessException, java.lang.reflect.InvocationTargetException, InstantiationException, java.beans.IntrospectionException, CTPException {
-      BusinessSampleUnit businessSampleUnit = processLine(columnPositionMappingStrategy, nextLine);
-      List<String> namesOfInvalidColumns = validateLine(businessSampleUnit);
-
-      // If a unit ref is already registered
-      if (unitRefs.contains(businessSampleUnit.getSampleUnitRef())) {
-        log.warn("This sample unit ref {} is duplicated in the file.", businessSampleUnit.getSampleUnitRef());
-        throw new CTPException(CTPException.Fault.VALIDATION_FAILED,
-                String.format("This sample unit ref %s is duplicated in the file.", businessSampleUnit.getSampleUnitRef()));
-      }
-      unitRefs.add(businessSampleUnit.getSampleUnitRef());
-
-          if (!namesOfInvalidColumns.isEmpty()) {
-            String errorMessage = String.format("Error in %s due to field(s) %s", Arrays.toString(nextLine),
-                    Arrays.toString(namesOfInvalidColumns.toArray()));
-            log.warn(errorMessage);
-            throw new CTPException(CTPException.Fault.VALIDATION_FAILED, errorMessage);
-          }
-          businessSampleUnit.setSampleUnitType("B");
-
-      return businessSampleUnit;
-  }
-
-  private List<String> validateLine(BusinessSampleUnit csvLine) {
-    Set<ConstraintViolation<BusinessSampleUnit>> violations = getValidator().validate(csvLine);
-    List<String> invalidFields = violations.stream().map(v -> v.getPropertyPath().toString())
+
+        return new ImmutablePair<>(sampleUnitList, errors);
+    }
+
+    public SampleSummary ingest(final SampleSummary sampleSummary, final MultipartFile file)
+            throws Exception {
+
+        int maxErrors = this.appConfig.getSampleIngest().getMaxErrors();
+
+        Pair<List<BusinessSampleUnit>, List<ValidationError>> result = parseAndValidateFile(file, maxErrors);
```

Comment:
The logic has gotten pretty gnarly. Would it be worth parsing and validating separately?

Reply:
Yeah this was just how it ended up merging the conflicts - you split the parse into a separate method which would probably have been done differently had the changes I made already been there. I'll revisit